
<h1 align="center"> Segmenting and Clustering Neighborhoods in Toronto </h1>

## PART 1. 
* **Build a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name in Toronto.** 

### Importing Libraries/Data

In [1]:
import pandas as pd
import requests

In [2]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Response 200 means that we are able to make the connection to the page

In [3]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [4]:
len(wiki_data), type(wiki_data)

(3, list)

In [5]:
## We need the first table alone, so dropping the other tables
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [6]:
## Dropping Borough which are not assigned
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [7]:
## Grouping the records based on Postal Code
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [8]:
## Checking for number of records where Neighbourhood is "Not assigned"
df.Neighbourhood.str.count("Not assigned").sum()

0

In [9]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [10]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [11]:
df.shape

(103, 3)

**Answer to Part 1: We have 103 rows and 3 columns**

## PART 2.
### Get the geographical coordinates of the neighborhoods in Toronto. Use the Geocoder package or the csv file to create dataframe with longitude and latitude values

In [15]:
# Installing geocoder
!pip install geocoder

In [16]:
import geocoder # import geocoder

In [17]:
# Retrieving LAT/LONG via geocoder didnt work 

# initialize your variable to None
# lat_lng_coords = None

#postal_code = 'M3A'

# loop until you get the coordinates
# while(lat_lng_coords is None):
#  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
#  lat_lng_coords = g.latlng

# latitude = lat_lng_coords[0]
# longitude = lat_lng_coords[1]


In [18]:
# so I followed the approach to import the CSV file from the URL as stated in the assignment notes

data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [19]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


In [20]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [21]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [22]:
# Joining postal codes
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [23]:
combined_data.shape

(103, 5)

**Part 2 verification:** We get 103 rows as expected when we do a inner join, so we have good data.

## PART 3. 
* **Explore and cluster the neighborhoods in Toronto, utilizing k-mean clustering and foursquare API** 

In [24]:
from geopy.geocoders import Nominatim 

In [25]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [26]:
pip install folium

Note: you may need to restart the kernel to use updated packages.


In [27]:
import folium

In [28]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

In [29]:
# Initializing Foursquare API credentials
CLIENT_ID = 'KDSWY0KDE4ZA114EGMOD0QFBFGUMT5Y2SWBLY11ZYFQMEHBB' 
CLIENT_SECRET = 'D5WENVTK3LGAI1DJPKX3S2IHM0S5EHVO4IXEBAT2Z1DTRXMU'
VERSION = '20180604' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: KDSWY0KDE4ZA114EGMOD0QFBFGUMT5Y2SWBLY11ZYFQMEHBB
CLIENT_SECRET:D5WENVTK3LGAI1DJPKX3S2IHM0S5EHVO4IXEBAT2Z1DTRXMU


In [30]:
# function definition to get all the venue categories in Toronto

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [31]:
# Collecting the venues in Toronto for each Neighbourhood
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [32]:
venues_in_toronto.shape

(1320, 5)

In [33]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1         Brookbanks Pool                   Pool  
2           Variety Store      Food & Drink Shop  
3  Victoria Village Arena           Hockey Arena  
4               Portugril  Portuguese Restaurant

In [34]:
# Checking the Venues based on Neighbourhood
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1305  Mimico NW, The Queensway West, South of Bloor,...   
1306  Mimico NW, The Queensway West, South of Bloor,...   
1307  Mimico NW, The Queensway West, South of Bloor,...   
1308  Mimico NW, The Queensway West, South of Bloor,...   
1309  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656         Brookbanks Pool   
2                  43.753259               -79.329656           Variety Store   
3                  43.725882               -79.315572  Victoria Village Arena   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1305               43.628841               -79.520999              Wingporium   
1306               43.628841               -79.520999        South St. Burger   
1307               43.628841               -79.520999               Dollarama   
1308               43.628841               -79.520999          Healthy Planet   
1309               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1                      Pool  
2         Food & Drink Shop  
3              Hockey Arena  
4     Portuguese Restaurant  
...                     ...  
1305            Wings Joint  
1306           Burger Joint  
1307         Discount Store  
1308        Supplement Shop  
1309                 Bakery  

[412 rows x 5 columns]

In [35]:
# Checking for the maximum venue categories
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.379817   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.321558   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                          The National Club  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[235 rows x 4 columns]

In [36]:
# One Hot encoding the venue Categories
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1315                  0        0                   0             0   
1316                  0        0                   0             0   
1317                  0        0                   0             0   
1318                  0        0                   0             0   
1319                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1315               0                0                 0                    0   
1316               0                0                 0                    0   
1317               0                0                 0                    0   
1318               0                0                 0                    0   
1319               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Truck Stop  Vegetarian / Vegan Restaurant  \
0                0         0  ...           0                              0   
1                0         0  ...           0                              0   
2                0         0  ...           0                              0   
3                0         0  ...           0                              0   
4                0         0  ...           0                              0   
...            ...       ...  ...         ...                            ...   
1315             0         0  ...           0                              0   
1316             0         0  ...           0                              0   
1317             0         0  ...           0                              0   
1318             0         0  ...           0                              0   
1319             0         0  ...           0                              0   

      Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                    0            0                      0                0   
1                    0            0                      0                0   
2                    0            0                      0                0   
3                    0            0                      0                0   
4                    0            0                      0                0   
...                ...          ...                    ...              ...   
1315                 0            0                      0                0   
1316                 0            0                      0                0   
1317                 0            0                      0                0   
1318                 0            0                      0                0   
1319                 0            0                      0                0   

      Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0            0            0              0            0  
1            0            0              0            0  
2            0            0              0            0  
3            0            0              0            0  
4   

In [37]:
# Adding the neighbourhood to the encoded dataframe
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2         Parkwoods                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Truck Stop  \
0                    0             0  ...           0   
1                    0             0  ...           0   
2                    0             0  ...           0   
3                    0             0  ...           0   
4                    0             0  ...           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 236 columns]

In [38]:
# We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood 
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Truck Stop  \
0               0.0             0.000000           0.0  ...         0.0   
1               0.0             0.000000           0.0  ...         0.0   
2               0.0             0.000000           0.0  ...         0.0   
3               0.0             0.000000           0.0  ...         0.0   
4               0.0             0.047619           0.0  ...         0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 236 columns]

In [39]:
# Function definition to get the top most common venue categories
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [40]:
import numpy as np

In [41]:
# Get the top 10 to cluster the neighbourhoods
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1                Coffee Shop        Sandwich Place                   Pub   
2                Coffee Shop              Pharmacy            Restaurant   
3                       Bank    Chinese Restaurant   Japanese Restaurant   
4             Sandwich Place    Italian Restaurant       Thai Restaurant   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0           Yoga Studio       Ethiopian Restaurant           Escape Room   
1          Skating Rink                        Gym             Drugstore   
2           Supermarket  Middle Eastern Restaurant         Deli / Bodega   
3           Yoga Studio                Event Space  Ethiopian Restaurant   
4              Pharmacy                Pizza Place                   Pub   

  8th Most Common Venue        9th Most Common Venue  \
0     Electronics Store  Eastern European Restaurant   
1            Donut Shop                      Dog Run   
2        Ice Cream Shop                  Bridal Shop   
3           Escape Room            Electronics Store   
4            Restaurant                         Café   

        10th Most Common Venue  
0                    Drugstore  
1                Deli / Bodega  
2           Chinese Restaurant  
3  Eastern European Restaurant  
4                      Butcher

### Clustering the Neighbourhoods

In [42]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [43]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [44]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 3,
       1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 3, 1, 1, 1, 1, 3, 4, 1, 1, 1, 1,
       3, 1, 3, 1, 1, 2, 0, 1, 0, 1, 3, 1, 1, 1, 4, 1, 1, 0, 1, 1, 1, 0,
       4, 4, 1, 3, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 3,
       3, 0, 1, 1, 1, 1, 3], dtype=int32)

In [45]:
# Adding the clustering Label column to the top 10 common venue categories
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [46]:
# Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Pool   
1  43.725882 -79.315572             1.0          Intersection   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                  Park     Food & Drink Shop            Yoga Studio   
1     French Restaurant           Coffee Shop  Portuguese Restaurant   
2                Bakery                  Park         Breakfast Spot   
3     Accessories Store           Coffee Shop               Boutique   
4      Sushi Restaurant        Sandwich Place     Italian Restaurant   

  5th Most Common Venue   6th Most Common Venue  7th Most Common Venue  \
0      Department Store             Escape Room      Electronics Store   
1          Hockey Arena     Distribution Center     Dim Sum Restaurant   
2                  Café             Yoga Studio      French Restaurant   
3           Event Space  Furniture / Home Store  Vietnamese Restaurant   
4   Japanese Restaurant                Beer Bar   Fast Food Restaurant   

         8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant              Drugstore             Donut Shop  
1                        Diner         Discount Store            Yoga Studio  
2                          Pub  Performing Arts Venue          Historic Site  
3           Dim Sum Restaurant             Comic Shop            Escape Room  
4                Smoothie Shop          Burrito Place    Distribution Center

In [47]:
# Drop all the NaN values to prevent data skew
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

In [48]:
# Plotting
import matplotlib.cm as cm
import matplotlib.colors as colors

In [49]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

## CLuster Verification 

### Cluster 1

In [50]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Pool   
21              York             0.0                  Park   
32       Scarborough             0.0            Playground   
35         East York             0.0                  Park   
52        North York             0.0                   Gym   
68   Central Toronto             0.0                  Park   
83   Central Toronto             0.0          Tennis Court   
85       Scarborough             0.0            Playground   
91  Downtown Toronto             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Park     Food & Drink Shop           Yoga Studio   
21         Women's Store                  Pool           Yoga Studio   
32           Pizza Place    College Rec Center       College Stadium   
35     Convenience Store          Intersection           Yoga Studio   
52                  Park          Dessert Shop           Event Space   
68      Sushi Restaurant         Jewelry Store                 Trail   
83            Playground                  Park                 Trail   
85                  Park          Intersection           Yoga Studio   
91            Playground                 Trail           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0       Department Store           Escape Room     Electronics Store   
21      Department Store           Escape Room     Electronics Store   
32  Ethiopian Restaurant           Escape Room     Electronics Store   
35          Dessert Shop  Ethiopian Restaurant           Escape Room   
52  Ethiopian Restaurant           Escape Room     Electronics Store   
68           Yoga Studio          Dessert Shop           Escape Room   
83        Discount Store      Department Store          Dessert Shop   
85      Department Store           Escape Room     Electronics Store   
91      Department Store           Escape Room     Electronics Store   

          8th Most Common Venue        9th Most Common Venue  \
0   Eastern European Restaurant                    Drugstore   
21  Eastern European Restaurant                    Drugstore   
32  Eastern European Restaurant                    Drugstore   
35            Electronics Store  Eastern European Restaurant   
52  Eastern European Restaurant                    Drugstore   
68            Electronics Store  Eastern European Restaurant   
83           Dim Sum Restaurant                        Diner   
85  Eastern European Restaurant                    Drugstore   
91  Eastern European Restaurant                    Drugstore   

   10th Most Common Venue  
0              Donut Shop  
21             Donut Shop  
32             Donut Shop  
35              Drugstore  
52             Donut Shop  
68              Drugstore  
83            Yoga Studio  
85             Donut Shop  
91             Donut Shop

### Cluster 2

In [51]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             1.0          Intersection   
2    Downtown Toronto             1.0           Coffee Shop   
3          North York             1.0        Clothing Store   
4    Downtown Toronto             1.0           Coffee Shop   
7          North York             1.0                   Gym   
..                ...             ...                   ...   
96   Downtown Toronto             1.0    Italian Restaurant   
97   Downtown Toronto             1.0            Restaurant   
99   Downtown Toronto             1.0           Coffee Shop   
100      East Toronto             1.0    Light Rail Station   
102         Etobicoke             1.0            Kids Store   

    2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1       French Restaurant           Coffee Shop  Portuguese Restaurant   
2                  Bakery                  Park         Breakfast Spot   
3       Accessories Store           Coffee Shop               Boutique   
4        Sushi Restaurant        Sandwich Place     Italian Restaurant   
7              Beer Store            Restaurant            Coffee Shop   
..                    ...                   ...                    ...   
96                   Café            Restaurant            Coffee Shop   
97                   Café           Coffee Shop     Seafood Restaurant   
99              Bookstore   Japanese Restaurant       Sushi Restaurant   
100           Yoga Studio            Skate Park          Auto Workshop   
102     Convenience Store         Tanning Salon                 Bakery   

      5th Most Common Venue   6th Most Common Venue   7th Most Common Venue  \
1              Hockey Arena     Distribution Center      Dim Sum Restaurant   
2                      Café             Yoga Studio       French Restaurant   
3               Event Space  Furniture / Home Store   Vietnamese Restaurant   
4       Japanese Restaurant                Beer Bar    Fast Food Restaurant   
7       Japanese Restaurant          Clothing Store      Italian Restaurant   
..                      ...                     ...                     ...   
96                   Bakery               Pet Store               Gift Shop   
97   Gluten-free Restaurant               Speakeasy                     Pub   
99        Indian Restaurant     Martial Arts School              Restaurant   
100                 Brewery           Burrito Place              Comic Shop   
102         Supplement Shop          Discount Store  Thrift / Vintage Store   

    8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
1                   Diner         Discount Store            Yoga Studio  
2                     Pub  Performing Arts Venue          Historic Site  
3      Dim Sum Restaurant             Comic Shop            Escape Room  
4           Smoothie Shop          Burrito Place    Distribution Center  
7             Supermarket         Discount Store                   Café  
..                    ...                    ...                    ...  
96             Beer Store                Butcher   Caribbean Restaurant  
97             Steakhouse             Food Court                 Bakery  
99                   Café               Beer Bar   Ethiopian Restaurant  
100        Farmers Market   Fast Food Restaurant                 Garden  
102        Sandwich Place           Burger Joint   Fast Food Restaurant  

[73 rows x 12 columns]

### Cluster 3

In [52]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             2.0  Fast Food Restaurant          Dessert Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6           Event Space  Ethiopian Restaurant           Escape Room   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
6     Electronics Store  Eastern European Restaurant             Drugstore   

  9th Most Common Venue 10th Most Common Venue  
6            Donut Shop                Dog Run

### Cluster 4

In [53]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
44      Scarborough             3.0              Bus Line   
49       North York             3.0                  Park   
50       North York             3.0                   Gym   
56             York             3.0        Sandwich Place   
61  Central Toronto             3.0                  Park   
63             York             3.0           Pizza Place   
64             York             3.0     Convenience Store   
66       North York             3.0                  Park   
67  Central Toronto             3.0  Gym / Fitness Center   
71      Scarborough             3.0           Auto Garage   
77        Etobicoke             3.0           Pizza Place   

         2nd Most Common Venue 3rd Most Common Venue  \
44                      Bakery          Soccer Field   
49  Construction & Landscaping                Bakery   
50                Intersection           Pizza Place   
56           Convenience Store                   Bar   
61                 Swim School              Bus Line   
63              Breakfast Spot     Convenience Store   
64               Jewelry Store           Yoga Studio   
66           Convenience Store                   Bar   
67              Breakfast Spot                   Gym   
71               Shopping Mall                Bakery   
77                        Park              Bus Line   

        4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
44              Metro Station          Intersection        Ice Cream Shop   
49                Yoga Studio    Dim Sum Restaurant           Event Space   
50                 Comic Shop                 Diner           Event Space   
56               Skating Rink        Discount Store           Yoga Studio   
61                Yoga Studio          Dessert Shop           Escape Room   
63         Dim Sum Restaurant           Event Space  Ethiopian Restaurant   
64         Dim Sum Restaurant           Event Space  Ethiopian Restaurant   
66                Yoga Studio    Dim Sum Restaurant           Event Space   
67                      Hotel     Food & Drink Shop      Department Store   
71  Middle Eastern Restaurant        Sandwich Place           Yoga Studio   
77             Sandwich Place      Department Store     Electronics Store   

          7th Most Common Venue        8th Most Common Venue  \
44                  Yoga Studio                        Diner   
49         Ethiopian Restaurant                  Escape Room   
50         Ethiopian Restaurant                  Escape Room   
56                  Escape Room            Electronics Store   
61            Electronics Store  Eastern European Restaurant   
63                  Escape Room            Electronics Store   
64                  Escape Room            Electronics Store   
66         Ethiopian Restaurant                  Escape Room   
67                 Dance Studio                         Park   
71                  Escape Room            Electronics Store   
77  Eastern European Restaurant                    Drugstore   

          9th Most Common Venue       10th Most Common Venue  
44                  Escape Room            Electronics Store  
49            Electronics Store  Eastern European Restaurant  
50            Electronics Store  Eastern European Restaurant  
56  Eastern European Restaurant                    Drugstore  
61                    Drugstore                   Donut Shop  
63  Eastern European Restaurant                    Drugstore  
64  Eastern European Restaurant                    Drugstore  
66            Electronics Store  Eastern European Restaurant  
67               Sandwich Place                  Pizza Place  
71  Eastern European Restaurant                    Drugstore  
77                   Donut Shop                      Dog Run

### Cluster 5

In [54]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
12       Scarborough             4.0                   Bar   
57        North York             4.0        Baseball Field   
62   Central Toronto             4.0          Home Service   
98         Etobicoke             4.0                  Pool   
101        Etobicoke             4.0                  Pool   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12           Home Service           Yoga Studio    Dim Sum Restaurant   
57            Yoga Studio        Farmers Market           Event Space   
62                 Garden           Yoga Studio      Department Store   
98                  River           Yoga Studio         Deli / Bodega   
101        Baseball Field           Yoga Studio          Dessert Shop   

    5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
12            Event Space  Ethiopian Restaurant                  Escape Room   
57   Ethiopian Restaurant           Escape Room            Electronics Store   
62   Ethiopian Restaurant           Escape Room            Electronics Store   
98            Escape Room     Electronics Store  Eastern European Restaurant   
101  Ethiopian Restaurant           Escape Room            Electronics Store   

           8th Most Common Venue        9th Most Common Venue  \
12             Electronics Store  Eastern European Restaurant   
57   Eastern European Restaurant                    Drugstore   
62   Eastern European Restaurant                    Drugstore   
98                     Drugstore                   Donut Shop   
101  Eastern European Restaurant                    Drugstore   

    10th Most Common Venue  
12               Drugstore  
57              Donut Shop  
62              Donut Shop  
98                 Dog Run  
101             Donut Shop

### END OF ASSIGNMENT 